radar-upload-backend/src/main/java/org/radarbase/upload/doa/RecordRepositoryImpl.kt

```diff
@@ -209,38 +232,64 @@ class RecordRepositoryImpl(@Context private var em: javax.inject.Provider<Entity
-        }
         logger.debug("Updating record $id status from ${existingMetadata.status} to ${metadata.status}")
         existingMetadata.apply {
-            status = RecordStatus.valueOf(metadata.status)
+            metadata.status?.let { newStatus ->
+                if (!allowedStateTransition(existingMetadata.status, newStatus)) {
+                    throw ConflictException("incompatible_status", "Record cannot be updated: Conflict in record meta-data status. " +
+                            "Cannot transition from state ${existingMetadata.status} to ${metadata.status}")
+                }
+
+                status = RecordStatus.valueOf(newStatus)
+            }
             message = metadata.message
         }.update()
 
         merge<RecordMetadata>(existingMetadata)
     }
 
-    override fun delete(record: Record) = em.get().transact { remove(record) }
+    override fun delete(record: Record, revision: Int) = em.get().transact {
+        refresh(record)
+        if (record.metadata.revision != revision) {
+            throw ConflictException("incompatible_revision", "Revision in metadata ${record.metadata.revision} does not match provided revision $revision")
+        }
+        if (record.metadata.status == RecordStatus.PROCESSING) {
+            throw ConflictException("incompatible_status", "Cannot delete record ${record.id}: it is currently being processed.")
+        }
+        remove(record)
+    }
 
     override fun readMetadata(id: Long): RecordMetadata? = em.get().transact { find(RecordMetadata::class.java, id) }
 
     companion object {
         private val logger = LoggerFactory.getLogger(RecordRepositoryImpl::class.java)
+
+        private fun allowedStateTransition(from: RecordStatus, to: String): Boolean {
+            val toStatus = try {
+                RecordStatus.valueOf(to)
+            } catch (ex: IllegalArgumentException) {
+                throw BadRequestException("unknown_status", "Record status $from is not a known status. Use one of: ${RecordStatus.values().joinToString()}.")
+            }
+
+            return toStatus == from
+                    || toStatus in allowedStateTransitions[from]
+                    ?: throw BadRequestException("unknown_status", "Record status $from is not a known status. Use one of: ${RecordStatus.values().joinToString()}.")
+        }
+
+        private val allowedStateTransitions: Map<RecordStatus, Set<RecordStatus>> = EnumMap<RecordStatus, Set<RecordStatus>>(RecordStatus::class.java).apply {
+            this[RecordStatus.INCOMPLETE] = setOf(RecordStatus.READY, RecordStatus.FAILED)
+            this[RecordStatus.READY] = setOf(RecordStatus.QUEUED, RecordStatus.FAILED)
+            this[RecordStatus.QUEUED] = setOf(RecordStatus.PROCESSING, RecordStatus.READY, RecordStatus.FAILED)
+            this[RecordStatus.PROCESSING] = setOf(RecordStatus.READY, RecordStatus.SUCCEEDED, RecordStatus.FAILED)
```

Comment:
If the record goes to PROCESSING, stage I would expect it to be either SUCCEEDED or FAILED. then allow FAILED record to be able to rest to READY again. This will allow us to make sure that a SUCCEEDED record doesn't need to be reprocessed and give full control to the connector to decide the FAILED and SUCCEEDED status.

Reply:
So my reasoning here was that there are multiple paths `READY` means that a record can be processed (whether it was processed or queried before would not matter, only if it has been uploaded to Kafka).

Scenarios where `PROCESSING` moves to `READY`:
-> timeout of the connector (e.g. connector was restarted while processing)
-> connection problems with kafka (connector converted the record but could not upload it)

In both cases, the connector should try to process them again at a later time.

In my opinion `FAILED` and `SUCCEEDED` are end states (non-recoverable). I could add a state like `TEMPORARY_FAILURE`, but I don't know what it would add to `READY` without adding some more metadata as well.

Reply:
ok. That makes sense. :+1:  I also do not think of having an explicit TEMPORARY_FAILURE state. We can stick to FAILED and SUCCEEDED as the end states and proceed with this flow.